### ISSUE TITLE
Support the use system cacerts as an option

### ISSUE THREAD

**Issue by gaeljw** [adoptium/containers#293]
### Please add the exact image (with tag) that you are using

eclipse-temurin:17-jdk-jammy

### Please add the version of Docker you are running

podman 4.2.1

### What happened?

Hello folks,

Not sure if it can be considered a bug or a feature.

I switched from the openjdk:17-bullseye image to eclipse-temurin:17-jdk-jammy and noticed that the JVM doesn't use system cacerts with eclipse-temurin image.

I'm adding a private ca cert as per Debian standards.

With openjdk image nothing more needed for the JVM to successfully connect to HTTPS servers referencing the private CA. But with eclipse-temurin, I'm getting errors `sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target`

I didn't fully investigate this yet but I think the reason might be that there was a special handling in openjdk images:
https://github.com/docker-library/openjdk/blob/master/Dockerfile-linux.template#L238
```
# update "cacerts" bundle to use Debian's CA certificates (and make sure it stays up-to-date with changes to Debian's store)
# see https://github.com/docker-library/openjdk/issues/327
#     http://rabexc.org/posts/certificates-not-working-java#comment-4099504075
#     https://salsa.debian.org/java-team/ca-certificates-java/blob/3e51a84e9104823319abeb31f880580e46f45a98/debian/jks-keystore.hook.in
#     https://git.alpinelinux.org/aports/tree/community/java-cacerts/APKBUILD?id=761af65f38b4570093461e6546dcf6b179d2b624#n29
	{ \
		echo '#!/usr/bin/env bash'; \
		echo 'set -Eeuo pipefail'; \
		echo 'trust extract --overwrite --format=java-cacerts --filter=ca-anchors --purpose=server-auth "$JAVA_HOME/lib/security/cacerts"'; \
	} > /etc/ca-certificates/update.d/docker-openjdk; \
	chmod +x /etc/ca-certificates/update.d/docker-openjdk; \
	/etc/ca-certificates/update.d/docker-openjdk; \
	\
# https://github.com/docker-library/openjdk/issues/331#issuecomment-498834472
	find "$JAVA_HOME/lib" -name '*.so' -exec dirname '{}' ';' | sort -u > /etc/ld.so.conf.d/docker-openjdk.conf; \
	ldconfig; \
```
(Relates to https://github.com/docker-library/openjdk/issues/327 as mentioned in the comments)

Any opinion on this?

I feel it should not be mandatory to customize the JVM truststore if the cacert is already trusted at system level. Meaning I would expect to just have to do `update-ca-certificates` in the eclipse-temurin image but nothing more.

**Comment by karianna** [comment]
@gaeljw The Docker image code you list there is not from the Adoptium project (our images are under eclipse-temurin).  Can you tell us where you grabbed the image from?

**Comment by gaeljw** [comment]
@karianna that's the point: using the https://hub.docker.com/_/eclipse-temurin image (version `17-jdk-jammy`) has a different behaviour than the one from Openjdk and I'm wondering if this is not a behaviour that should/could be implemented in the eclipse-temurin image.

**Comment by karianna** [comment]
@gaeljw This is going to take sometime for us to untangle in terms of a long term decision.

In the short term we'd recommend symlinking to the system cacerts file in order to use those instead of the cacerts file shipped with Temurin.

**Comment by gaeljw** [comment]
Thanks for the heads-up @karianna , I totally understand this is not an easy decision and already worked around it on my side :)

**Comment by gaeljw** [comment]
>  can I confirm that you are building a downstream image, and injecting your CA into the system CA store, using something like (...)

Yes, absolutely.

> Do you run any processes in the resulting container other than java?

Most of the time, only Java.

> One could instead substitute the COPY/RUN commands above for ones that operated on the JKS keystore directly. (Perhaps: COPY to /tmp/something, RUN keytool -import…).

True for sure.

> Reasons I can think of to prefer it in the system CA store are (...)

It's mainly option 3 (you have a bunch of other, non-java containers that you also add the certificate to, and want to use the same approach across them all) in my case.

And also a kind of "habit" because on regular VMs we don't modify the JKS keystore, thus I was somehow hoping to keep the same habit.

> For Adoptium, a complication for adopting an approach like docker/openjdk used was it's system-specific, in this case Debian and derivatives, and Adoptium build on top of CentOS and Alpine (and Windows) too, and a different solution is needed in each case.

Absolutely make sense. If I were you, I would have probably came to the same conclusion to be fair.

> Another approach is to configure your container runtime host to have the correct CAs and then readonly-mount the trust store into every container you run.

Indeed, this approach works for "runtime" container (although it requires some commands, we can't pass the root CA as is, it has to be imported with keytool, isn't it?).

It's less convenient for "build" containers (think of running maven image for instance on a developer laptop) where having to mount stuff degrades the experience IMHO.

> What container engine are you using?

Podman or docker on laptops. Containers through Kubernetes elsewhere.

-----

I appreciate the detailed answer anyway, thanks a lot.

**Comment by rassie** [comment]
I would like to add my two cents, since I haven't expected the migration from `openjdk` to `eclipse-temurin` to be bumpy and have been bitten by this issue.

> Hi @gaeljw, can I confirm that you are building a downstream image, and injecting your CA into the system CA store, using something like
> 
> ```
> FROM <openjdk image>
> COPY my-ca.crt /usr/local/share/ca-certificates
> RUN update-ca-certificates
> ```

In my case, I don't. It's my job to make externally developed applications deployable to a customer's data center, which includes configuring an outgoing HTTP proxy. This particular proxy terminates TLS connections corporate-style, so that any certificate must be validated with the corporate CA certificate. With `openjdk`-based containers, I've been mapping a volume with those CA certificates to `/usr/local/share/ca-certificates` and added `update-ca-certificates` to the container command.

> Do you run any processes in the resulting container other than `java`?
> 
> The reason I ask is, if there's nothing other than `java` running in the container, there's no reason to put the certificate into the OS trust store. One could instead substitute the `COPY`/`RUN` commands above for ones that operated on the JKS keystore directly. (Perhaps: `COPY` to `/tmp/something`, `RUN keytool -import…`)

This is certainly possible, but requires awareness from development and ops teams. Do I understand correctly that this process would change again with 18+, because of https://bugs.openjdk.org/browse/JDK-8275252?

> Reasons I can think of to prefer it in the system CA store are
> 
> 1. you are running processes other than java in the container
> 2. you are (more) familiar with adding to the system CA store
> 3. you have a bunch of other, non-java containers that you also add the certificate to, and want to use the same approach across them all

For me it's a mixture of all three. In general, I consider Java a hassle ops-wise, because of the tendency to avoid OS-typical paths and to create own solutions, so that even something as seemingly simple as adding a proxy becomes a daunting multi-day task (because as an ops engineer you need to find out what actually works for a combination of JRE, framework and particular application implementation. Depending on your luck, not even using a combination of system properties, environment variables and added files will help you). While I'm not explicitely running several processes in a container, there is a chance that some other tool called from the application might need the system store, so there is no way around using both.

> For Adoptium, a complication for adopting an approach like docker/openjdk used was it's system-specific, in this case Debian and derivatives, and Adoptium build on top of CentOS and Alpine (and Windows) too, and a different solution is needed in each case.

I would consider adding CA certificates to the container store a task common enough to warrant explicit support by the image, i.e.  there should be a pre-defined directory for additional certificates and the entrypoint should take care of adding those certificates to the system store and Java store. The particular implementation would of course be different for different base OSs, but I think there is merit to keeping a correct implementation in the image instead of making each user implement it themselves. 

> Another approach is to configure your container runtime host to have the correct CAs and then readonly-mount the trust store into every container you run. What container engine are you using?

This will become infeasible with Kubernetes, since there would be no way to install anything on the host system (especially with immutable container-only systems).

**Comment by rassie** [comment]
Addenum: would you accept a patch adding some sort of support? I think I could manage to provide an implementation for Linux-based containers, someone else would need to add the Windows part.

**Comment by karianna** [comment]
> Addenum: would you accept a patch adding some sort of support? I think I could manage to provide an implementation for Linux-based containers, someone else would need to add the Windows part.

Are you suggesting a patch to this repo or something hosted indpendently?

**Comment by rassie** [comment]
> > Addenum: would you accept a patch adding some sort of support? I think I could manage to provide an implementation for Linux-based containers, someone else would need to add the Windows part.
> 
> Are you suggesting a patch to this repo or something hosted indpendently?

If I understand the project structure correctly, then a PR on this repository. Something that will be in `eclipse-temurin` image after being merged.

**Comment by karianna** [comment]
Will bring up at the PMC meeting for discussion, thanks!

**Comment by jerboaa** [comment]
> > > Addenum: would you accept a patch adding some sort of support? I think I could manage to provide an implementation for Linux-based containers, someone else would need to add the Windows part.
> > 
> > 
> > Are you suggesting a patch to this repo or something hosted indpendently?
> 
> If I understand the project structure correctly, then a PR on this repository. Something that will be in `eclipse-temurin` image after being merged.

@rassie Please explain the design that you have in mind. In particular, how use-case a) use temurin cacerts only and use-case b) use system cacerts over temurin cacerts both work with the same image. That would help us make a better decision on the PMC. Personally, I don't see a good way to support both `a)` and `b)` with the same image.

**Comment by rassie** [comment]
> @rassie Please explain the design that you have in mind. In particular, how use-case a) use temurin cacerts only and use-case b) use system cacerts over temurin cacerts both work with the same image. That would help us make a better decision on the PMC. Personally, I don't see a good way to support both `a)` and `b)` with the same image, but I might be missing something.

From what I understand, use-case `a)` is basically the status quo, we wouldn't need to change anything to support it. Use-case `b)` is from my point of view is covered by the design previously deployed by `openjdk`, which automatically deployed all system CA certificates to the keystore, updating/overwriting CA certificates in the keystore. There would be a use-case `c)`, which would be using system CA certs for certs not already available via Temurin keystore, but I don't think there is much demand for that.

If I'm not missing something glaring, implementing `a)` and `b)` in one image requires a simple option for activating `b)` (which incidentally is what this issue is basically about, having an option). I'm confident that introducing an environment variable along the lines of `TEMURIN_USE_SYSTEM_CA` would be enough for most people in this thread. The implementation itself would be very similar to the previous implementation from `openjdk`, with adjustments made for Alpine/CentOS specifics.

**Comment by jerboaa** [comment]
OK. So the "support system CA" feature would throw in extra authorities found in the system ca database into the JDK cacerts keystore. Got it. That's indeed a new use-case, `c)` use both, temurin cacerts and system ca certs combined. I take it, `a)` and `c)` would be sufficient? `b)` would be to use system certs **only**.

**Comment by rassie** [comment]
>  I take it, a) and c) would be sufficient? b) would be to use system certs only.

From my point of view, absolutely. I just need to add some CAs and make them visible to the JRE. The others (@gaeljw?) can pitch in with their use cases.

**Comment by tellison** [comment]
I'm going to reopen this while the discussion continues.

**Comment by gaeljw** [comment]
My use case would be to add some CAs and make them visible to the JRE and other tools like Git or cURL if they are part of the image. I don't see a use case where I would need to ignore some CAs. A "cumulative" approach where a bunch of CAs are available even if I don't need them is fine for me.

> I'm confident that introducing an environment variable along the lines of TEMURIN_USE_SYSTEM_CA would be enough for most people in this thread

@rassie , this would be an environment variable to pass at runtime? Or an `ARG` when building image? Not sure to grasp the idea here.

**Comment by rassie** [comment]
> > I'm confident that introducing an environment variable along the lines of TEMURIN_USE_SYSTEM_CA would be enough for most people in this thread
> @rassie , this would be an environment variable to pass at runtime? Or an `ARG` when building image? Not sure to grasp the idea here.

A runtime environment variable, so that you could use the current image as-is and if you needed to add system CA certs to the trust store, you'd add a volume and an environment variable to enable an integration script.

**Comment by gaeljw** [comment]
Okay, then I will define another use case I have : I'm building container images for building Maven/sbt projects in my company which uses a private registry with specific CA. For this use case, running such a container is done in context where I don't want to mount a volume or define environment variable. The image should "by default" contain the certificates and be configured to use them.

**Comment by jmtd** [comment]
> Okay, then I will define another use case I have : I'm building container images for building Maven/sbt projects in my company which uses a private registry with specific CA. For this use case, running such a container is done in context where I don't want to mount a volume or define environment variable. The image should "by default" contain the certificates and be configured to use them.

In this scenario, could you, as a separate process, build an internal image, downstream of this one, with the custom CA trust store? In which case, you could attack the problem at build time instead of runtime.

**Comment by gaeljw** [comment]
> In this scenario, could you, as a separate process, build an internal image, downstream of this one, with the custom CA trust store? 

That's exactly what I do currently.

I'm can live with that but with OpenJDK image I just had to download the CA and put it in the system certificates, now I have to do more stuff to also include it in the JVM trust store. (At build time I mean).

Basically I just copy pasted what OpenJDK image was doing 😅

------

For runtime usages, the env variable should be okay.

Actually we could provide the env variable at build time and just add the file mount at runtime, right?

**Comment by jerboaa** [comment]
@rassie Please explain how you plan to implement option `c.)` as you've mentioned in https://github.com/adoptium/containers/issues/293#issuecomment-1340960640 that use case `a.)` (status quo) and `c.)` are sufficient for you and we haven't heard any other requests. In particular how you'd support both use cases in the same image (some high level design). Once we know that and we are confident it doesn't break the common case `a.)` we should have all the info for a PMC decision. Then I'll bring it to the PMC for a decision and if approved your patch could go in.

**Comment by jerboaa** [comment]
> * For Red Hat/CentOS/ubi images, `update-ca-trust` already updates an external Java truststore (which would need to be symlinked to get activated). `update-ca-trust` would need to be patched at image build time to add the conditional to also update the truststore in `JAVA_HOME`.

This seems a no-go to me. `update-ca-trust` is part of the `ca-certificates` RPM package and we should not change it. This breaks security updates to the package. An alternative solution seems to be to make the Temurin JDK install use `/etc/pki/java/cacerts` instead somehow. By first adding certs from the JDK to the source dir where `update-ca-trust` looks. Not sure if that's feasible, though.

**Comment by jerboaa** [comment]
> * For Debian/Ubuntu/Alpine images, reintroduce the `update-ca-certificates` update hook and extend it with a conditional to only proceed if it's set.

Do you have a link to how that works?

**Comment by rassie** [comment]
> > * For Red Hat/CentOS/ubi images, `update-ca-trust` already updates an external Java truststore (which would need to be symlinked to get activated). `update-ca-trust` would need to be patched at image build time to add the conditional to also update the truststore in `JAVA_HOME`.
> 
> This seems a no-go to me. `update-ca-trust` is part of the `ca-certificates` RPM package and we should not change it. This breaks security updates to the package. An alternative solution seems to be to make the Temurin JDK install use `/etc/pki/java/cacerts` instead somehow. By first adding certs from the JDK to the source dir where `update-ca-trust` looks. Not sure if that's feasible, though.

Yes, this has been a pain point in my thought train too. Old `openjdk` images replaced `$JAVA_HOME/lib/security/cacerts` with a symlink to `/etc/pki/ca-trust/extracted/java/cacerts` at build time (which is essentially what you are proposing), so running `update-ca-trust` updated JRE truststore unconditionally. If we want to add a conditional (which we do), we can't do that at build time when the symlinks are created, so we need *some* way to affect the runtime. Patching `update-ca-trust` seems like an easier way, the alternative would be moving the symlink creation to the entrypoint script (which would have to be created first).

> This breaks security updates to the package.

How common are security updates inside a running container right now? I would have thought this is a non-issue, since most images get rebuilt on security updates, but maybe that's my personal bubble. 

But I've still checked :) I've got a rather old `centos:7` image lying around (`centos@sha256:be65f488b7764ad3638f236b7b515b3678369a5124c47b8d32916d6487418ea4`). I've patched `update-ca-trust` inside it in a primitive way and run `yum update ca-certificates`. `/usr/bin/update-ca-trust` has been replaced fully without an error message or even a warning.

**Comment by rassie** [comment]
> > * For Debian/Ubuntu/Alpine images, reintroduce the `update-ca-certificates` update hook and extend it with a conditional to only proceed if it's set.
> 
> Do you have a link to how that works?

The original script is quoted in the [first post](https://github.com/adoptium/containers/issues/293#issue-1415512239) of this issue. I would have extended it with something like `if [ -z "$TEMURIN_IMPORT_SYSTEM_CA" ]; then exit 0; fi` right after the shebang.

**Comment by jerboaa** [comment]
> > > * For Red Hat/CentOS/ubi images, `update-ca-trust` already updates an external Java truststore (which would need to be symlinked to get activated). `update-ca-trust` would need to be patched at image build time to add the conditional to also update the truststore in `JAVA_HOME`.
> > 
> > 
> > This seems a no-go to me. `update-ca-trust` is part of the `ca-certificates` RPM package and we should not change it. This breaks security updates to the package. An alternative solution seems to be to make the Temurin JDK install use `/etc/pki/java/cacerts` instead somehow. By first adding certs from the JDK to the source dir where `update-ca-trust` looks. Not sure if that's feasible, though.
> 
> Yes, this has been a pain point in my thought train too. Old `openjdk` images replaced `$JAVA_HOME/lib/security/cacerts` with a symlink to `/etc/pki/ca-trust/extracted/java/cacerts` at build time (which is essentially what you are proposing), so running `update-ca-trust` updated JRE truststore unconditionally. If we want to add a conditional (which we do), we can't do that at build time when the symlinks are created, so we need _some_ way to affect the runtime. Patching `update-ca-trust` seems like an easier way, the alternative would be moving the symlink creation to the entrypoint script (which would have to be created first).

The approach of exporting JDK certs to a source dir of where `update-ca-trust` expects them, running `update-ca-trust` and replacing the `cacerts` file with a symlink from an entrypoint script seems more amenable to me. This would only run iff the user sets some env and it doesn't change integrity of files owned by the `update-ca-trust` package.

> > This breaks security updates to the package.

Let me rephrase. The key point should be for `rpm -V ca-certificates` to pass for the temurin images based on RPM distros.
 
> How common are security updates inside a running container right now? I would have thought this is a non-issue, since most images get rebuilt on security updates, but maybe that's my personal bubble.

That's true, but patching something which isn't in our control risks breaking the auto-rebuild, does it not? E.g. if it changes in a way which breaks the patching mechanism.

**Comment by jerboaa** [comment]
> > The approach of exporting JDK certs to a source dir of where `update-ca-trust` expects them, running `update-ca-trust` and replacing the `cacerts` file with a symlink from an entrypoint script seems more amenable to me. This would only run iff the user sets some env and it doesn't change integrity of files owned by the `update-ca-trust` package.
> 
> I think we can agree on that, we can iron out the particular details in the PR, if and when it comes to that. Do we have all bases covered, i.e. do you have enough information for the PMC decision?

Yes, thanks.

**Comment by jerboaa** [comment]
@rassie Sorry for the delay.

> i.e. do you have enough information for the PMC decision?

We've discussed this issue on the PMC meeting last week and there were no objections to get the proposed solution implemented. Please be sure to add automated tests ensuring this works as expected for all scenarios.

**Comment by rassie** [comment]
@jerboaa Thanks!

> Please be sure to add automated tests ensuring this works as expected for all scenarios.

This sounds like a lot of fun. Since I'm new to the repository, I'll probably need a lot of feedback -- but I'll make an initial PR first and then we can iterate from there.

**Comment by jerboaa** [comment]
> This sounds like a lot of fun. Since I'm new to the repository, I'll probably need a lot of feedback -- but I'll make an initial PR first and then we can iterate from there.

Sounds good.

**Comment by jerboaa** [comment]
For reference the PMC decision about this will show up here once the May 3, 2023 minutes show up here:
https://github.com/adoptium/adoptium/tree/main/Meetings/PMC%20Meeting%20Minutes/2023

**Comment by rassie** [comment]
@jerboaa @gaeljw I've created a (draft) PR (#392), would appreciate your feedback.

**Comment by gdams** [comment]
We've merged the proposed changes - as a precautionary measure we're going to update the JDK20 manifests first for a week or two to see if anyone complains and then we'll go for a full rollout

**Comment by gdams** [comment]
We've had no bugs reported so a PR has been raised to update the remaining versions: https://github.com/docker-library/official-images/pull/15162
